rest/src/main/java/discord4j/rest/http/client/ClientRequest.java

```diff
@@ -17,41 +17,92 @@
 
 package discord4j.rest.http.client;
 
+import discord4j.rest.request.DiscordRequest;
+import discord4j.rest.route.Route;
+import discord4j.rest.util.RouteUtils;
+import io.netty.handler.codec.http.DefaultHttpHeaders;
 import io.netty.handler.codec.http.HttpHeaderNames;
 import io.netty.handler.codec.http.HttpHeaders;
 import io.netty.handler.codec.http.HttpMethod;
 
+import java.util.Optional;
+
 /**
- * A wrapper over a partial HTTP client request definition.
+ * An adapted request definition from an original {@link DiscordRequest}.
  */
 public class ClientRequest {
 
-    private final HttpMethod method;
+    private final DiscordRequest<?> request;
     private final String url;
     private final HttpHeaders headers;
 
-    public ClientRequest(HttpMethod method, String url, HttpHeaders headers) {
-        this.method = method;
-        this.url = url;
-        this.headers = headers;
+    /**
+     * Create a new {@link ClientRequest} from the given request template.
+     *
+     * @param request the {@link DiscordRequest} template
+     */
+    public ClientRequest(DiscordRequest<?> request) {
+        this.request = request;
+        this.url = RouteUtils.expandQuery(request.getCompleteUri(), request.getQueryParams());
+        this.headers = Optional.ofNullable(request.getHeaders())
+                .map(map -> map.entrySet().stream()
+                        .reduce((HttpHeaders) new DefaultHttpHeaders(), (headers, entry) -> {
```

Comment:
Unfortunately netty still forces a cast due to `add` method returning `HttpHeaders`. It would look like this:
```java
this.headers = Optional.ofNullable(request.getHeaders())
		.map(map -> map.entrySet().stream()
				.reduce(new DefaultHttpHeaders(), (headers, entry) -> {
					String key = entry.getKey();
					entry.getValue().forEach(value -> headers.add(key, value));
					return headers;
				}, (e1, e2) -> (DefaultHttpHeaders) e1.add(e2)))
		.orElse(new DefaultHttpHeaders());
```